File: java/erdosnumbers.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class erdosnumbers<T> {
    public static void main(String[] args) {
        erdosnumbers e = new erdosnumbers();
        e.run();
    }

    HashMap<String, Integer> shortcut = new HashMap<>();
    public void run() {
        Scanner sc = new Scanner(System.in);
        HashMap<String, HashSet<String>> graf = new HashMap<>();
        HashSet<String> seen = new HashSet<>();
        HashSet<String> seenStart = new HashSet<>();
        ArrayList<String> startNodes = new ArrayList<>();
        while (sc.hasNextLine()) {
            String input = sc.nextLine();
            if(input.isEmpty()) break;
            String[] s = input.split(" ");
            String startNode = s[0];
            startNodes.add(startNode);
            seenStart.add(startNode);
            for (String name : s) {
                seen.add(startNode);
                if(!graf.containsKey(startNode)) graf.put(startNode, new HashSet<>());
                if(!graf.containsKey(name)) graf.put(name, new HashSet<>());
                HashSet<String> hs = graf.get(startNode);
                hs.add(name); 
                hs = graf.get(name);
                hs.add(startNode);
            }
        }
        sc.close();

        for (String name : seen) {
            if(!seenStart.contains(name)) {
                graf.remove(name);
            }
        }

        String endNode = "PAUL_ERDOS";
        shortcut.put(endNode, 0);
        bfs(graf, endNode, "bla");
        for (String name : startNodes) {
            if (shortcut.containsKey(name)) System.out.println(name + " " + shortcut.get(name));
            else System.out.println(name + " no-connection");
        }
    }

    private ArrayList<String> bfs(HashMap<String, HashSet<String>> graf, String startNode, String endNode) {
        Queue<String> possibleRouts = new LinkedList<>();
        HashMap<String, String> pastNodes = new HashMap<>();
        ArrayList<String> path = new ArrayList<>();
        if(startNode != endNode) possibleRouts.add(startNode);

        pastNodes.put(startNode, startNode);
        outer: while(!possibleRouts.isEmpty()){
            String v = possibleRouts.poll();
            for(String s : graf.get(v)){
                if(!pastNodes.containsKey(s)){
                    possibleRouts.add(s);
                    pastNodes.put(s, v);
                    shortcut.put(s, shortcut.get(v)+1);
                    if(s.equals(endNode)){
                        while(true){
                            path.add(s);
                            if(s.equals(startNode)) break outer;
                            s = pastNodes.get(s);
                        }
                    }
                }
            }
        }

        Collections.reverse(path);
        return path;
    }
}

File: rust/src/bin/erdosnumbers.rs
use std::{collections::{HashMap, HashSet, VecDeque}, io::{self, BufRead}};

fn main() {
    let input = read_input();

    let mut start_nodes = Vec::new();
    let mut graph: HashMap<String, HashSet<String>> = HashMap::new();
    for line in input {
        start_nodes.push(line.get(0).unwrap().clone());
        if !graph.contains_key(line.get(0).unwrap()) {
            graph.insert(line.get(0).unwrap().clone(), HashSet::new());
        }
        for (i, name) in line.iter().enumerate() {
            if i == 0 {continue;}
            let hs = graph.get_mut(line.get(0).unwrap()).unwrap();
            hs.insert(name.clone());
            if let Some(hs2) = graph.get_mut(name) {
                hs2.insert(start_nodes.last().unwrap().clone());
            } else {
                graph.insert(name.clone(), HashSet::from([start_nodes.last().unwrap().clone()]));
            }
        }
    }

    let end_node = "PAUL_ERDOS".to_string();
    let (_, hm) = bfs(&graph, end_node.clone(), "bla".to_string());
    for st in start_nodes {
        if st == end_node {
            println!("{} 0", st);
        } else if let Some(n) = hm.get(&st) {
            println!("{} {}", st, *n)
        } else {
            println!("{} no-connection", st)
        }
    }

}

fn bfs2(graph: &HashMap<i32, HashSet<i32>>, start_node: i32, end_node: i32) -> Vec<i32> {
    let mut previous_node = HashMap::new();
    let mut queue = VecDeque::from(vec![start_node]);
    let mut path = Vec::new();
    if start_node == end_node {
        return vec![end_node];
    }

    'outer: while let Some(current_node) = queue.pop_front() {
        for mut neighbor in graph.get(&current_node).unwrap() {
            if !previous_node.contains_key(neighbor) {
                previous_node.insert(neighbor, current_node);
                queue.push_back(*neighbor);
                if *neighbor == end_node {
                    path.push(*neighbor);
                    while *neighbor != start_node {
                        neighbor = previous_node.get(neighbor).unwrap();
                        path.push(*neighbor);
                    }
                    break 'outer;
                }
            }
        }
    }

    path.reverse();
    path
}

fn bfs(graph: &HashMap<String, HashSet<String>>, start_node: String, end_node: String) -> (Vec<String>, HashMap<String, i32>){
    let mut previous_node = HashMap::from([(start_node.clone(),start_node.clone())]);
    let mut queue = VecDeque::from(vec![start_node.clone()]);
    let mut path = Vec::new();
    let mut distances = HashMap::from([(start_node.clone(), 0)]);
    if start_node == end_node {
        return (vec![end_node.to_string()], distances);
    }

    'outer: while let Some(current_node) = queue.pop_front() {
        for mut neighbor in graph.get(&current_node).unwrap() {
            if !previous_node.contains_key(neighbor) {
                previous_node.insert(neighbor.clone(), current_node.clone());
                queue.push_back(neighbor.clone());
                distances.insert(neighbor.clone(), distances.get(&current_node).unwrap()+1);
                if *neighbor == end_node {
                    path.push(neighbor.to_string());
                    while *neighbor != start_node {
                        neighbor = previous_node.get(neighbor).unwrap();
                        path.push(neighbor.to_string());
                    }
                    break 'outer;
                }
            }
        }
    }

    path.reverse();
    (path, distances)
}

fn read_input() -> Vec<Vec<String>> {
    let mut lines = vec![];
    let stdin = io::stdin();

    for line in stdin.lock().lines() {
        match line {
            Ok(input) => {
                if input.trim().is_empty() {break;}
                lines.push(input.trim().split(" ").map(|s| s.to_string()).collect())
            }
            Err(_) => {
                break;
            }
        }
    }

    lines
}